docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+the scoreboard in memory in the Job Controller. Make JobStatus more succinct,
+and makes for fewer.shorted JobStatus updates.  Client needs to list all pods of the
+job, and extract the one that has the right index (using the annotation).
+
+#### Job Spec Compatilibity
+
+A job spec written before this change will work exactly the same
+as before.  The Pods it creates will have the same environment as before.
+
+However, if the job controller version is reverted, to a version before this change,
+the jobs whose pod specs depend on the the new annotation will fail.  This is
+okay for a Beta resource.
+
+#### Job Controller Changes
+
+When Job controller sees that the number of running pods is less than the desired
+parallelism of the job, it finds the first index in the scoreboard with value
+`notRunning`.  It creates a pod with this creation index.
+
+
+When it creates a pod with creation index `i`,  it makes a copy
+of the `.spec.template`, and sets
+`.spec.template.metadata.annotations.[kubernetes.io/job/completion-index]`
+to `i`.   It does this in both the index-only and multiple-substitutions options.
+
+
+In the multiple substitutions option, for each `KEY` in map
+`PerCompletionEnvVars`, the job controller appends an EnvVar
+to the Pod it creates, like:
+
+```
+Pod.Spec.container[*].EnvVars.name=KEY
+Pod.Spec.container[*].EnvVars.value=PerCompletionEnvVars[KEY][COMPLETION_INDEX]
+```
+
+Then it creates the pod.
+
+
+When the controller notices that a pod has completed or is running or failed,
+it updates the scoreboard.
+
+
+When all entries in the scoreboard are `complete`, then the job is complete.
+
+**TBD**: whether it also POSTs scoreboard to the JobStatus.
+See [this
+comment](https://github.com/kubernetes/kubernetes/pull/18016#discussion_r46615718),
+Brian Grant says not to store a list of pod statuses in the PetSetStatus.
+Presumably this applies to JobStatus too.  Need to reconcile these.
+Client, such as kubectl can list all pods of the job and use a field selector
+on annotations and then pick out the pod(s) with the right index.
+
+
+#### Downward API Changes
+
+No changes are needed to the downward API.
+
+#### Kubectl Interface Changes
+
+The `--completions` and `--completion-index-var-name` flags are added to kubectl.
+
+For example, this command:
+
+```
+kubectl run say-number --image=busybox \
+   --completions=3 \
+   --completion-index-var-name=I \
+   -- \
+   sh -c 'echo "My index is $I" && sleep 5' 
+```
+
+will run 3 pods to completion, each printing one of the following lines:
+
+```
+My index is 1
+My index is 2
+My index is 0
+```
+
+Kubectl will also support the `--per-completion-env` flag, as described previously.
+For example, this command:
+
+```
+kubectl run say-fruit --image=busybox \
+   --per-completion-env=FRUIT="apple banana cherry" \
+   --per-completion-env=COLOR="green yellow red" \
+   -- \
+   sh -c 'echo "Have a nice $COLOR $FRUIT" && sleep 5' 
```

Comment:
Since we support env bar expansion inside the args and cmd, having an example that leverages that would be good (since you don't have to do shell magic).  Maybe just as an API only example